File: 2020/5/디스크 컨트롤러.py
import heapq

def solution(jobs):
    answer, time, end, count = 0, 0, -1, 0
    hq = []
    n= len(jobs)
    while count < n:
        for job in jobs:
            if end < job[0] <= time:
                heapq.heappush(hq, job[1])
                answer += time - job[0]
        if hq:
            answer += hq[0] * len(hq)
            end = time
            time += heapq.heappop(hq)
            count += 1
        else:
            time+=1
            
    return int(answer/n)

File: 2020/11/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int time =0; int start = -1; int count = 0; int answer = 0;
        int n = jobs.length;
        
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        while(count < n){
            for(int[] job : jobs){
                if(start< job[0] && job[0] <= time){
                    pq.offer(job[1]); //작업 추가
                    answer += time - job[0]; //기다린시간 더하기
                }
            }
            if(!pq.isEmpty()){
                answer += pq.peek() * pq.size(); //작업 + 작업 대기시간 추가
                count +=1; //작업 완료
                start = time; //작업 시작 시간 갱신
                time += pq.poll(); //현재 시간 갱신
            }else{
                time++;
            }
        }
        return answer/n;
    }
}